modules/generative-zhipuai/clients/zhipuai.go

```diff
@@ -0,0 +1,408 @@
+)
+
+var compile, _ = regexp.Compile(`{([\w\s]*?)}`)
+
+func buildUrlFn(baseURL string) (string, error) {
+	 
+	path := "chat/completions"
+	 
+	return url.JoinPath(baseURL, path)
+}
+
+type zhipuai struct {
+	zhipuAIApiKey       string 
+	buildUrl           func(baseURL string) (string, error)
+	httpClient         *http.Client
+	logger             logrus.FieldLogger
+}
+
+func New(zhipuAIApiKey string, timeout time.Duration, logger logrus.FieldLogger) *zhipuai {
+	return &zhipuai{
+		zhipuAIApiKey:       zhipuAIApiKey, 
+		httpClient: &http.Client{
+			Timeout: timeout,
+		},
+		buildUrl: buildUrlFn,
+		logger:   logger,
+	}
+}
+
+func (v *zhipuai) GenerateSingleResult(ctx context.Context, textProperties map[string]string, prompt string, options interface{}, debug bool, cfg moduletools.ClassConfig) (*modulecapabilities.GenerateResponse, error) {
+	forPrompt, err := v.generateForPrompt(textProperties, prompt)
+	if err != nil {
+		return nil, err
+	}
+	return v.Generate(ctx, cfg, forPrompt, options, debug)
+}
+
+func (v *zhipuai) GenerateAllResults(ctx context.Context, textProperties []map[string]string, task string, options interface{}, debug bool, cfg moduletools.ClassConfig) (*modulecapabilities.GenerateResponse, error) {
+	forTask, err := v.generatePromptForTask(textProperties, task)
+	if err != nil {
+		return nil, err
+	}
+	return v.Generate(ctx, cfg, forTask, options, debug)
+}
+
+func (v *zhipuai) Generate(ctx context.Context, cfg moduletools.ClassConfig, prompt string, options interface{}, debug bool) (*modulecapabilities.GenerateResponse, error) {
+	params := v.getParameters(cfg, options) 
+	fmt.Println(params)
+	debugInformation := v.getDebugInformation(debug, prompt)
+
+	zpiUrl, err := v.buildZhipuAIUrl(ctx, params)
+	if err != nil {
+		return nil, errors.Wrap(err, "url join path")
+	}
+
+	input, err := v.generateInput(prompt, params)
+	if err != nil {
+		return nil, errors.Wrap(err, "generate input")
+	}
+	// print input
+	fmt.Println(input)
+	body, err := json.Marshal(input)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal body")
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", zpiUrl,
+		bytes.NewReader(body))
+	if err != nil {
+		return nil, errors.Wrap(err, "create POST request")
+	}
+	apiKey, err := v.getApiKey(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "ZhipuAI API Key")
+	}
+	req.Header.Add(v.getApiKeyHeaderAndValue(apiKey))
+ 
+	req.Header.Add("Content-Type", "application/json")
+
+	res, err := v.httpClient.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "send POST request")
+	}
+	defer res.Body.Close()
+
+	requestID := res.Header.Get("x-request-id")
+	bodyBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "read response body")
+	}
+
+	var resBody generateResponse
+	if err := json.Unmarshal(bodyBytes, &resBody); err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("unmarshal response body. Got: %v", string(bodyBytes)))
```

Comment:
we are not returning the response json in case of errors in our modules, so here we can just wrap the error